waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +642,145 @@
+               :instance-scheduling-start-times (vec instance-scheduling-start-times'')
+               :starting-instance-ids (vec starting-instance-ids'))))))
+
+(defn start-launch-metrics-maintainer
+  "go block to collect metrics on the instance launch overhead of waiter services.
+
+  Updated state for the router is read from `router-state-chan`."
+  [router-state-chan]
+  (let [exit-chan (async/chan 1)
+        query-chan (au/latest-chan)
+        update-state-timer (metrics/waiter-timer "state" "launch-metrics-maintainer" "update-state")
+        waiter-schedule-timer (metrics/waiter-timer "launch-overhead" "schedule-time")]
+    (async/go
+      (try
+        (loop [{:keys [known-service-ids service-id->launch-tracker] :as current-state}
+               {:known-service-ids #{}
+                :service-id->launch-tracker {}}]
+          (let [new-state
+                (async/alt!
+                  exit-chan
+                  ([message]
+                   (if (= :exit message)
+                     (do
+                       (log/warn "stopping launch-metrics-maintainer")
+                       (comment "Return nil to exit the loop"))
+                     current-state))
+
+                  router-state-chan
+                  ([router-state]
+                   (timers/start-stop-time!
+                     update-state-timer
+                     (let [{:keys [service-id->healthy-instances service-id->instance-counts service-id->instance-slots
+                                   service-id->my-instance->slots service-id->unhealthy-instances]} router-state
+                           incoming-service-ids (set (keys service-id->my-instance->slots))
+                           new-service-ids (set/difference incoming-service-ids known-service-ids)
+                           removed-service-ids (set/difference known-service-ids incoming-service-ids)
+                           service-id->launch-tracker' (update-launch-trackers
+                                                         service-id->launch-tracker new-service-ids removed-service-ids
+                                                         service-id->healthy-instances service-id->unhealthy-instances
+                                                         service-id->instance-counts waiter-schedule-timer)]
+                       {:known-service-ids incoming-service-ids
+                        :service-id->launch-tracker service-id->launch-tracker'})))
+
+                  query-chan
+                  ([{:keys [cid response-chan]}]
+                   (cid/cinfo cid (str "returning current launch state " current-state))
+                   (let [sanitized-state
+                         (->>
+                           #(dissoc % :service-schedule-timer :service-startup-timer)
+                           (partial pc/map-vals)
+                           (update-in current-state [:service-id->launch-tracker]))]
```

Comment:
Nitpick: this threading macro was hard for me to parse. This would have been easier to read:
```
(update-in 
  current-state [:service-id->launch-tracker]
  (fn [service-id->launch-tracker]
    (pc/map-vals
      #(dissoc % :service-schedule-timer :service-startup-timer)
      service-id->launch-tracker)))
```

Reply:
De-threaded.